mesonbuild/msubprojects.py

```diff
@@ -312,6 +312,61 @@ def foreach(r, wrap, repo_dir, options):
     mlog.log(out, end='')
     return True
 
+def purge(r: Resolver, wrap: PackageDefinition, repo_dir: str, options: argparse.Namespace) -> bool:
+    # if subproject is not wrap-based, then don't remove it
+    if not wrap.type:
+        return True
+
+    if wrap.type == 'redirect':
+        redirect_file = Path(wrap.filename).resolve()
+        if options.confirm:
+            redirect_file.unlink()
+        mlog.log(f'Deleting {redirect_file}')
+
+    if options.include_cache:
+        packagecache = Path(r.cachedir).resolve()
+        subproject_cache_file = packagecache / wrap.get("source_filename")
```

Comment:
I've hit an error: `ERROR: Missing key 'source_filename' in pygobject.wrap`. The problem is here, `source_filename` is optional key in the wrap file. wrap-git usually don't have it.

Reply:
I see just below you have a try block to catch WrapException, you need the same for this block. I gave that a try locally and it works.